mllib/src/main/scala/org/apache/spark/mllib/clustering/package.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib
+
+import breeze.linalg.{ Vector => BV }
+
+import org.apache.spark.mllib.linalg.Vector
+import org.apache.spark.rdd.RDD
+
+package object base {
```

Comment:
I'm willing to take one last crack at contributing the Bregman modifications to Spark. Here is an layering of the pull requests:

1) modifications to BLAS to support arbitrary functions and axpy on SparseVectors
2) definition of PointOps trait and various implementations for different divergences 
3) rewrite of clusterer to use PointOps trait but only using the squared Euclidean distance implementation 
4) generalization of clustering interface to select different divergences.
5) store seeds
6) maintain exactly K clusters by splitting clusters heuristicly.

I can submit PRs for 1-4. 

Thoughts?

Sent from my iPhone

> On Jan 8, 2015, at 4:00 PM, Xiangrui Meng <notifications@github.com> wrote:
> 
> In mllib/src/main/scala/org/apache/spark/mllib/clustering/package.scala:
> 
> > + *
> > + * Unless required by applicable law or agreed to in writing, software
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.mllib
> > +
> > +import breeze.linalg.{ Vector => BV }
> > +
> > +import org.apache.spark.mllib.linalg.Vector
> > +import org.apache.spark.rdd.RDD
> > +
> > +package object base {
> Though private, having some doc would make the code easy to understand. For example,
> 
> What does FP mean?
> inh?
> ...
> —
> Reply to this email directly or view it on GitHub.
>